src/session/client-certificate.c

```diff
@@ -49,31 +93,55 @@
  * including "0::" prefix and newline.
  * NB: the kernel doesn't allow newlines in cgroup names. */
 static char *
-read_proc_self_cgroup (size_t *out_length)
+read_proc_pid_cgroup (int dirfd, size_t *out_length)
 {
-  FILE *fp = fopen ("/proc/self/cgroup", "r");
-
-  if (fp == NULL)
-    {
-      warn ("Failed to open /proc/self/cgroup");
-      return NULL;
-    }
-
   char buffer[1024];
-  *out_length = fread (buffer, 1, sizeof (buffer) - 1, fp);
-  buffer[*out_length] = '\0';
-  fclose (fp);
+  ssize_t len = read_proc_file (dirfd, "cgroup", buffer, sizeof buffer);
+  if (len < 0)
+      return NULL;
 
-  if (*out_length > 5 &&  /* at least "0::/\n" */
-      *out_length <= sizeof (buffer) - 1 && /* not too big */
+  if (len > 5 &&  /* at least "0::/\n" */
+      len <= sizeof (buffer) - 1 && /* not too big */
       strncmp (buffer, "0::", 3) == 0 && /* must be a cgroupsv2 */
-      buffer[*out_length - 1] == '\n') /* must end with a newline */
-    return strdupx (buffer);
+      buffer[len - 1] == '\n') /* must end with a newline */
+    {
+      *out_length = len;
+      return strdupx (buffer);
+    }
 
   warnx ("unexpected cgroups content, certificate matching only supports cgroup v2: '%s'", buffer);
```

Comment:
I wonder if you should unconditionally bail out in this case.  This is unexpected, right?  We only go down this path if we're on a certificate authentication, and in that case, we had better be able to read the cgroup.  The only thing I can think of is that the login process went away while we were in the middle of doing this, and then the correct thing to do is to exit.  We have `errx()` in the "get start time" code which would also be hit in that case.

My point: if we remove any attempt at seeming like we handle errors gracefully then we can also simplify error handling in `read_proc_file()` and switch it over too `err()` instead of `warn()`, changing its return type to `size_t` instead of `ssize_t`.

Reply:
I suppose this depends on how hard we want cockpit to fail. Right now the code ends up in `exit_init_problem()` to report a proper protocol error, and could possibly fail over to another authentication mechanism. Of course it's not super useful to enable cert auth on a cgroupsv1 system, but I don't feel confident enough to just ruin error reporting that way (especially since I've already spent several hours on *improving* error reporting for various corner cases :sweat_smile: )

Reply:
I did that in #21292. Let's review that first, and once we find agreement to the approach and land it, I'll rebase this one and apply the same thing to the other exit paths here.

Reply:
Rebased already, couldn't resist. The error handling looks much nicer/more robust now indeed!

Reply:
This should all be much more robust now.